source/postprocess/visualization/named_additional_outputs.cc

```diff
@@ -0,0 +1,172 @@
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/postprocess/visualization/named_additional_outputs.h>
+#include <aspect/material_model/interface.h>
+#include <aspect/utilities.h>
+
+#include <algorithm>
+
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      template <int dim>
+      NamedAdditionalOutputs<dim>::
+      NamedAdditionalOutputs ()
+        :
+        DataPostprocessor<dim> ()
+      {}
+
+
+      template <int dim>
+      void
+      NamedAdditionalOutputs<dim>::
+      initialize ()
+      {
+        MaterialModel::MaterialModelOutputs<dim> out(0,
+                                                     this->n_compositional_fields());
+        this->get_material_model().create_additional_named_outputs(out);
+
+        for (unsigned int k=0; k<out.additional_outputs.size(); ++k)
+          {
+            const MaterialModel::NamedAdditionalMaterialOutputs<dim> *result
+              = dynamic_cast<const MaterialModel::NamedAdditionalMaterialOutputs<dim> *> (out.additional_outputs[k].get());
+
+            if (result)
+              {
+                std::vector<std::string> names = result->get_names();
+
+                for (unsigned int i=0; i<names.size(); ++i)
+                  property_names.push_back(names[i]);
+              }
+          }
+
+        for (unsigned int i=0; i<property_names.size(); ++i)
+          std::replace(property_names[i].begin(),property_names[i].end(),' ', '_');
+      }
+
+      template <int dim>
+      std::vector<std::string>
+      NamedAdditionalOutputs<dim>::
+      get_names () const
+      {
+        return property_names;
+      }
+
+      template <int dim>
+      std::vector<DataComponentInterpretation::DataComponentInterpretation>
+      NamedAdditionalOutputs<dim>::
+      get_data_component_interpretation () const
+      {
+        std::vector<DataComponentInterpretation::DataComponentInterpretation> interpretation (get_names().size(),
+            DataComponentInterpretation::component_is_scalar);
+        return interpretation;
+      }
+
+      template <int dim>
+      UpdateFlags
+      NamedAdditionalOutputs<dim>::
+      get_needed_update_flags () const
+      {
+        return update_gradients | update_values  | update_q_points;
```

Comment:
hm, that's awkward -- we don't really know whether we need all of these flags. but I guess there is nothing we can do about that.